### ISSUE TITLE
Issue in the Bash terminal on `rocker/binder`: warning `pyenv` command not found

### ISSUE THREAD

**Issue by vemonet** [rocker-org/rocker-versioned2#428]
Hi, thanks a lot for making RStudio easily available on JupyterHub and Binder!

I ran the image using the command given on DockerHub (my laptop config is Fedora Linux 35 - kernel Linux 5.16.20-200.fc35.x86_64 - Docker version 20.10.14, build a224086):

```bash
docker run -p 8888:8888 rocker/binder
```

But I noticed 2 things that seems off:

1. This one is not a big problem, but there is a file `=2.0` that is created in the home folder, containing the output of a `pip install`:

```bash
Collecting jupyter-rsession-proxy
  Downloading jupyter_rsession_proxy-2.0.1-py3-none-any.whl (5.3 kB)
Collecting notebook
  Downloading notebook-6.4.10-py3-none-any.whl (9.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.9/9.9 MB 126.0 MB/s eta 0:00:00
```

2. When going to the terminal there is a warning displayed before the command input: 

```bash
bash: pyenv: command not found
rstudio@5020993aeee1:~$ 
```

I checked how to install `pyenv`, but, among all available tools for managing python environment, this one does not seems to be really developed: https://github.com/pyenv/pyenv#installation unfortunately they only provide installers for windows and macOS, but for linux systems there is not much so I am not sure it is used and tested a lot in linux environments

Is it normal that we are seeing this warning message about `pyenv` and the pip install output file?

How can we remove the warning message about `pyenv`? (preferably without installing it, to avoid having too many confusing ways to handle [python environments](https://xkcd.com/1987))

**Comment by eitsupi** [comment]
Thank you for investigating.
I agree that the Python-related installation scripts, used to `rocker/binder`, need improvement.

https://github.com/rocker-org/rocker-versioned2/blob/a97bd9e6538a946e6029fcfd371d78feb9a26eaa/scripts/install_python.sh
https://github.com/rocker-org/rocker-versioned2/blob/a97bd9e6538a946e6029fcfd371d78feb9a26eaa/scripts/install_pyenv.sh

However, I have not been able to start because of my lack of Python setup knowledge...
(I don't use `venv` or `pyenv` or other tools because I usually work inside a Docker container and use a different image for each project).

**Comment by cboettig** [comment]
Thanks @vemonet for reporting.  Will have to investigate why pyenv is not installing. 

For context, I agree with your comments about there being too many ways of managing python environments.  We introduced `pyenv` only because it used by `pipenv` to handle switching between Python versions.  Almost all of the many ways of managing python virtual environments I had looked at do not include an easy way to say "give me a environment that uses python 3.7'.  Notably, the default python with Ubuntu 20.04 base image we use is at python 3.8, and some modules like those that were built against tensorflow 1.x series never extended compatibility beyond python 3.7...  Still, we might be better off going the route that `reticulate` takes to install python versions through miniconda...

**Comment by eddelbuettel** [comment]
BioConductor have an entire package for that:  https://www.bioconductor.org/packages/release/bioc/html/basilisk.html

I mostly use the Ostrich algorithm and keep my head in the sand by not using Python much, or relying on what my (generally current) Ubuntu release gives me.

**Comment by vemonet** [comment]
Thanks @eitsupi and @cboettig for the quick help

I investigated a bit more and it seems like `pyenv` is properly installed in `/opt/pyenv`! The problem comes from the `$PATH` environment variable. It seems to be overwritten by something (probably RStudio or Texlive)

If I try to add `/opt/pyenv/bin` to the path in the Dockerfile:

```dockerfile
FROM rocker/binder:latest
USER ${NB_USER}
ENV PATH=$PATH:/opt/pyenv/bin
```

It is not in the path when I access RStudio after building and running the image:

```bash
rstudio@2e9f50e27756:~$ echo $PATH
/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
```

Any idea what would cause the `$PATH` to be overwritten? Maybe I should use a different way to set it in the `Dockerfile`

**Comment by eitsupi** [comment]
When the RStudio server is started by the `/init` command, the environment variables of the root user are written to the `Renviron.site` file so that they can be referenced in the sessions from RStudio. (#151)
However, RStudio sessions launched from Jupyter cannot use this workaround, so users would need to write environment variables to Renviron by hand now.

**Comment by vemonet** [comment]
Even when defining the PATH environment variable in `Renviron.site` it still does not apply to the environment variables when RStudio runs

Checking the `Renviron.site` file properly sets `PATH`:
```bash
rstudio@786149aa8c5f:~$ cat ${R_HOME}/etc/Renviron.site
R_LIBS=${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library'}
WORKON_HOME=/opt/venv
RETICULATE_MINICONDA_ENABLED=FALSE
PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux:/opt/pyenv/bin
```

But the path env variable is still unchanged:

```bash
bash: pyenv: command not found
rstudio@786149aa8c5f:~$ echo $PATH
/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
```

**Comment by eitsupi** [comment]
Is it possible that RStudio may be reading other Renviron file?
Are there any Renviron files in `rstudio`'s home directory or elsewhere?

**Comment by vemonet** [comment]
Note: I am not sure if [`install_pyenv.sh`](https://github.com/rocker-org/rocker-versioned2/blob/a97bd9e6538a946e6029fcfd371d78feb9a26eaa/scripts/install_pyenv.sh) is used, but the `~/.bashrc` file does not contains any change to the PATH. I tried to add a change to the PATH in `~/.bashrc`, just in case, and it does not help


>  Are there any Renviron files in rstudio's home directory or elsewhere?

It seems to be the only one in the container: 

```bash
rstudio@5e998beacbad:~$ find / -name "Renviron.site"
find: ‘/etc/ssl/private’: Permission denied
find: ‘/proc/tty/driver’: Permission denied
find: ‘/root’: Permission denied
find: ‘/run/sudo’: Permission denied
find: ‘/tmp/RtmpJPDyEE’: Permission denied
find: ‘/tmp/quarto-session01ec5a1b’: Permission denied
/usr/local/lib/R/etc/Renviron.site
find: ‘/var/cache/apt/archives/partial’: Permission denied
find: ‘/var/cache/ldconfig’: Permission denied
find: ‘/var/cache/private’: Permission denied
find: ‘/var/lib/private’: Permission denied
find: ‘/var/log/private’: Permission denied
```

Quite mysterious

**Comment by eitsupi** [comment]
As a workaround for now, the following Dockerfile could be used.

```Dockerfile
FROM rocker/binder:latest
USER root
ENV PATH=PATH=/opt/pyenv/bin:~/.local/bin:$PATH
RUN echo "PATH=${PATH}" >>${R_HOME}/etc/Renviron.site
USER ${NB_USER}
```

**Comment by vemonet** [comment]
Mmmh, unfortunately it does not fix it, I tried this exact Dockerfile

The `$PATH` env variable is still the same without the additions:

```bash
echo $PATH
/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
```

If I check the `Renviron.site` the last PATH defined is the right one with `pyenv/bin`:

```bash
cat /usr/local/lib/R/etc/Renviron.site
R_LIBS=${R_LIBS-'/usr/local/lib/R/site-library:/usr/local/lib/R/library'}
PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux:/usr/local/texlive/bin/x86_64-linux/

WORKON_HOME=/opt/venv
RETICULATE_MINICONDA_ENABLED=FALSE
PATH=/opt/pyenv/bin:~/.local/bin:/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
PATH=/opt/pyenv/bin:~/.local/bin:/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
```

So it seems that the `Renviron.site` file is not picked up properly by the `entrypoint` script, and this entrypoint is completely overwritting the PATH environment variable just to put RStudio and texlive bin path

**Comment by vemonet** [comment]
If the rserver command, that seems to be run with the user rstudio does not pick up `Renviron.site` , then the path is probably overwritten by another script readable by the `rstudio` user

Let's search in the user home folder which scripts could potentially overwrite the path:

```bash
grep -rni ~/ -e "PATH="
/home/rstudio/.profile:21:    PATH="$HOME/bin:$PATH"
/home/rstudio/.profile:26:    PATH="$HOME/.local/bin:$PATH"
/home/rstudio/.profile:28:export PATH=/usr/lib/rstudio-server/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/texlive/bin/x86_64-linux
```

There we go! The RStudio environment variables are defined in the `~/.profile` !

And it was actually next to the `Renviron.site` command all along: https://github.com/rocker-org/rocker-versioned2/blob/master/scripts/install_binder.sh#L21 

**💡 This issue can be fixed by changing the hardcoded PATH in `.profile` at the end of your Dockerfile**

```dockerfile
RUN sed -i 's/^export PATH=.*$/export PATH=${PATH}/g' ~/.profile
```

Alternatively you can also add one more path export with `echo "PATH=${PATH} >> ~/.profile`, but it will create a long list of `export PATH`

Thanks for your help @eitsupi ! I'll make a pull request to fix the pyenv file in the binder deployment

**Comment by eitsupi** [comment]
@vemonet Thanks for digging that!
The build will run when R 4.2.0 will be released tomorrow, and I would like to merge this PR together with it.

**Comment by eitsupi** [comment]
Closed by #429 and #438
